### PR TITLE
Adding Decompression Registrar entry to the Streamer Stack

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
Streamer will process request from the DecompressionRegistrar before processing request from the FullFileDecompressor

Moved the creation of the DecompressorRegistrar from CompressionModule class to the CompressionModuleInterface class to allow it to be available in tooling application such as the Editor.

Also moved the registration of the Client CompressionSystemComponent to the CompressionModuleInterface to also allow it to be available with the Editor.

## How was this PR tested?

Enabled the Compression Gem and validated that the DecompressorRegistrar entry was created when running the AutomatedTesting.GameLauncher without issue.

### CODE REVIEW

**Review by lawsonamzn — APPROVED** [review]
looks boilerplate but correct, to me.

File: Gems/Compression/Code/Source/Clients/Streamer/DecompressorStackEntry.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        AZ_CLASS_ALLOCATOR_DECL;
+
+        ~DecompressorRegistrarConfig() override = default;
+        AZStd::shared_ptr<AZ::IO::StreamStackEntry> AddStreamStackEntry(
+            const AZ::IO::HardwareInformation& hardware, AZStd::shared_ptr<AZ::IO::StreamStackEntry> parent) override;
+        static void Reflect(AZ::ReflectContext* context);
+
+        //! Maximum number of reads that are kept in flight.
+        AZ::u32 m_maxNumReads{ 2 };
+        //! Maximum number of decompression task that can run simultaneously.
```
Comment (mbalfour-amzn) [inline Gems/Compression/Code/Source/Clients/Streamer/DecompressorStackEntry.h:55]:
```suggestion
        //! Maximum number of decompression tasks that can run simultaneously.
```

File: Gems/Compression/Code/Source/Clients/Streamer/DecompressorStackEntry.cpp [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+                "number will only be helpful if decompressing is faster than reading, otherwise the number of slots can be kept around the "
+                "same number as there are decompression slots."));
+            statistics.push_back(AZ::IO::Statistic::CreateInteger(
+                m_name, "Pending decompression", m_numPendingDecompression,
+                "The number of requests that have completed reading and are waiting for a decompression slot to become available. If this "
+                "value is frequently more than zero than the number of decompression slots may need to be increased, a faster decompressor "
+                "is needed or the number of read slots can be reduced."));
+            statistics.push_back(AZ::IO::Statistic::CreateByteSize(
+                m_name, "Buffer memory", m_memoryUsage,
+                "The total amount of memory in megabytes used by the decompressor. This is depended on the compressed file sizes and may "
```
Comment (mbalfour-amzn) [inline Gems/Compression/Code/Source/Clients/Streamer/DecompressorStackEntry.cpp:332]:
```suggestion
                "The total amount of memory in megabytes used by the decompressor. This is dependent on the compressed file sizes and may "
```

File: Gems/Compression/Code/Source/CompressionModuleInterface.cpp [hunk 3]
```diff
@@ -9,9 +9,12 @@
 #include "CompressionModuleInterface.h"
 #include <AzCore/Memory/Memory.h>
 
+#include <Clients/CompressionSystemComponent.h>
+
 #include <Compression/CompressionTypeIds.h>
+#include <Compression/DecompressionInterfaceAPI.h>
 
-#include <Clients/CompressionSystemComponent.h>
+#include <Clients/DecompressionRegistrarImpl.h>
```
Comment (spham-amzn) [inline Gems/Compression/Code/Source/CompressionModuleInterface.cpp:17]: nit: Can we group this include together with the `#include <Clients/CompressionSystemComponent.h>` above?

File: Gems/Compression/Code/Source/CompressionModuleInterface.cpp [hunk 4]
```diff
@@ -24,7 +27,22 @@ namespace Compression
     {
         m_descriptors.insert(m_descriptors.end(), {
             CompressionSystemComponent::CreateDescriptor(),
-            });
+        });
+
+        // Create and Register the Decompression Registrar
```
Comment (spham-amzn) [inline Gems/Compression/Code/Source/CompressionModuleInterface.cpp:32]:
```suggestion
        // Create and register the Decompression Registrar
```

File: Gems/Compression/Code/Source/Clients/Streamer/DecompressorStackEntry.cpp [hunk 5]
```diff
@@ -0,0 +1,810 @@
...
+                m_readBufferStatus[i] = ReadBufferStatus::ReadInFlight;
+
+                AZ_Assert(m_numInFlightReads < m_maxNumReads,
+                    "A FileRequest was queued for reading in DecompressorRegistrarEntry, but there's no slots available.");
+                m_numInFlightReads++;
+
+                return;
+            }
+        }
+        AZ_Assert(false, "%u of %u read slots are use in the DecompressorRegistrarEntry, but no empty slot was found.", m_numInFlightReads, m_maxNumReads);
```
Comment (mbalfour-amzn) [inline Gems/Compression/Code/Source/Clients/Streamer/DecompressorStackEntry.cpp:556]:
```suggestion
        AZ_Assert(false, "%u of %u read slots are used in the DecompressorRegistrarEntry, but no empty slot was found.", m_numInFlightReads, m_maxNumReads);
```

**Review by spham-amzn — APPROVED** [review]

File: Gems/Compression/Code/Source/CompressionModuleInterface.cpp (hunk 3 above)
Comment (lumberyard-employee-dm) [reply Gems/Compression/Code/Source/CompressionModuleInterface.cpp:17]: Yes we can

File: Gems/Compression/Code/Source/Clients/Streamer/DecompressorStackEntry.cpp [hunk 6]
```diff
@@ -0,0 +1,810 @@
...
+
+                        AZ::TaskToken token = taskGraph.AddTask(taskDescriptor, AZStd::move(decompressTask));
+                        token.Precedes(finishToken);
+                    }
+
+                    --m_numPendingDecompression;
+                    ++m_numRunningTasks;
+                }
+
+                AZ_Assert(m_taskGraphEvent->IsSignaled() == false, "Decompression has been started on another thread. while executing this function");
```
Comment (mbalfour-amzn) [inline Gems/Compression/Code/Source/Clients/Streamer/DecompressorStackEntry.cpp:685]:
```suggestion
                AZ_Assert(m_taskGraphEvent->IsSignaled() == false, "Decompression has been started on another thread while executing this function.");
```

**Review by mbalfour-amzn — APPROVED** [review]

File: Gems/Compression/Registry/decompressor_streamer.game.setreg [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+                    "Generic":
+                    {
+                        "Stack":
+                        {
+                            "Decompressor Registrar":
+                            {
+                                "$type": "DecompressorRegistrarConfig",
+                                // Slots after the "Decompressor" entry from the
+                                // `<engine-root>/Registry/streamer.game.setreg` file
+                                "$stack_after": "Decompressor",
```
Comment (AMZN-alexpete) [inline Gems/Compression/Registry/decompressor_streamer.game.setreg:19]: Interesting, wasn't aware of this ordering feature

**Pull request merged**